libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp

```diff
@@ -0,0 +1,356 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  CRSF protocol decoder based on betaflight implementation
+  Code by Andy Piper
+ */
+
+#include "AP_RCProtocol.h"
+#include "AP_RCProtocol_SRXL.h"
+#include "AP_RCProtocol_CRSF.h"
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_Math/crc.h>
+#include <AP_Vehicle/AP_Vehicle_Type.h>
+#include <AP_RCTelemetry/AP_CRSF_Telem.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+
+/*
+ * CRSF protocol
+ *
+ * CRSF protocol uses a single wire half duplex uart connection.
+ * The master sends one frame every 4ms and the slave replies between two frames from the master.
+ *
+ * 420000 baud
+ * not inverted
+ * 8 Bit
+ * 1 Stop bit
+ * Big endian
+ * 416666 bit/s = 46667 byte/s (including stop bit) = 21.43us per byte
+ * Max frame size is 64 bytes
+ * A 64 byte frame plus 1 sync byte can be transmitted in 1393 microseconds.
+ *
+ * CRSF_TIME_NEEDED_PER_FRAME_US is set conservatively at 1500 microseconds
+ *
+ * Every frame has the structure:
+ * <Device address><Frame length><Type><Payload><CRC>
+ *
+ * Device address: (uint8_t)
+ * Frame length:   length in  bytes including Type (uint8_t)
+ * Type:           (uint8_t)
+ * CRC:            (uint8_t)
+ *
+ */
+
+extern const AP_HAL::HAL& hal;
+
+// #define CRSF_DEBUG
+#ifdef CRSF_DEBUG
+# define debug(fmt, args...)	hal.console->printf("CRSF: " fmt "\n", ##args)
+static const char* get_frame_type(uint8_t byte)
+{
+    switch(byte) {
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_GPS:
+        return "GPS";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_BATTERY_SENSOR:
+        return "BATTERY";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_HEARTBEAT:
+        return "HEARTBEAT";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_VTX:
+        return "VTX";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_VTX_TELEM:
+        return "VTX_TELEM";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_PARAM_DEVICE_PING:
+        return "PING";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_COMMAND:
+        return "COMMAND";
+    default:
+        return "UNKNOWN";
```

Comment:
Is there any excuse for someone adding a frametype to not add a string here?  Adding "default:" means the compiler won't cause them pain until they conform.